File: content/tools/drill_gen.py
#!/usr/bin/env python3
"""Procedural problem generator for the Kumon-style speed & accuracy drills.

Pure arithmetic, no LLM — a level's "gen" spec (from drill_levels.json)
deterministically defines a family of problems; this module draws concrete,
non-duplicate problems from that family. Used by both the printable
worksheet builder (Python) and, eventually, the in-app drill (ported to Dart
with the same digit-range/carry rules so problems feel identical).

A "level" never mixes more than one new idea — mirrors the Kumon philosophy
of single-step increments (see the mind-map / study-guide precedent in this
repo for how we ground content decisions before building on them).
"""
import random


def digit_range(n):
    return {1: (1, 9), 2: (10, 99), 3: (100, 999), 4: (1000, 9999)}[n]


def _has_carry(a, b, width):
    da = [int(c) for c in str(a).zfill(width)]
    db = [int(c) for c in str(b).zfill(width)]
    return any(x + y >= 10 for x, y in zip(da, db))


def _has_borrow(a, b, width):
    da = [int(c) for c in str(a).zfill(width)]
    db = [int(c) for c in str(b).zfill(width)]
    return any(x < y for x, y in zip(da, db))


def _one(gen, rng):
    t = gen["type"]

    if t == "add":
        if "a" in gen and "b" in gen:
            for _ in range(500):
                a, b = rng.randint(*gen["a"]), rng.randint(*gen["b"])
                s = a + b
                if s < gen.get("min_sum", 0):
                    continue
                if s > gen.get("max_sum", 10 ** 9):
                    continue
                return {"key": ("add", a, b), "stem": f"{a} + {b} =", "answer": str(s)}
        ra, rb = digit_range(gen["digits_a"]), digit_range(gen["digits_b"])
        width = max(gen["digits_a"], gen["digits_b"])
        want_carry = gen.get("carry")
        for _ in range(500):
            a, b = rng.randint(*ra), rng.randint(*rb)
            if want_carry is not None and _has_carry(a, b, width) != want_carry:
                continue
            return {"key": ("add", a, b), "stem": f"{a} + {b} =", "answer": str(a + b)}

    if t == "add_column":
        rng_d = digit_range(gen["digits"])
        nums = [rng.randint(*rng_d) for _ in range(gen["count"])]
        return {"key": ("addcol", tuple(nums)), "stem": " + ".join(map(str, nums)) + " =",
                "answer": str(sum(nums))}

    if t == "sub":
        ra, rb = digit_range(gen["digits_a"]), digit_range(gen["digits_b"])
        width = max(gen["digits_a"], gen["digits_b"])
        want_borrow = gen.get("borrow")
        need_zero = gen.get("zeros", False)
        for _ in range(800):
            a, b = rng.randint(*ra), rng.randint(*rb)
            if a < b:
                continue
            if want_borrow is not None and _has_borrow(a, b, width) != want_borrow:
                continue
            if need_zero and "0" not in str(a)[1:]:
                continue
            return {"key": ("sub", a, b), "stem": f"{a} − {b} =", "answer": str(a - b)}

    if t == "table":
        table, lo_hi = gen["table"], gen["range"]
        n = rng.randint(*lo_hi)
        a, b = (table, n) if rng.random() < 0.5 else (n, table)
        return {"key": ("mul", table, n), "stem": f"{a} × {b} =", "answer": str(table * n)}

    if t == "table_mixed":
        table = rng.choice(gen["tables"])
        n = rng.randint(*gen["range"])
        a, b = (table, n) if rng.random() < 0.5 else (n, table)
        return {"key": ("mul", table, n), "stem": f"{a} × {b} =", "answer": str(table * n)}

    if t == "mul":
        ra, rb = digit_range(gen["digits_a"]), digit_range(gen["digits_b"])
        want_carry = gen.get("carry")
        for _ in range(500):
            a, b = rng.randint(*ra), rng.randint(*rb)
            if gen["digits_b"] == 1 and want_carry is not None:
                has_carry = any(int(d) * b >= 10 for d in str(a))
                if has_carry != want_carry:
                    continue
            return {"key": ("mul", a, b), "stem": f"{a} × {b} =", "answer": str(a * b)}

    if t == "div_fact":
        lo, hi = gen["range"]
        divisor, quotient = rng.randint(lo, hi), rng.randint(lo, hi)
        dividend = divisor * quotient
        return {"key": ("div", dividend, divisor), "stem": f"{dividend} ÷ {divisor} =",
                "answer": str(quotient)}

    if t == "div":
        lo, hi = digit_range(gen["digits_dividend"])
        dlo, dhi = gen["divisor"]
        for _ in range(500):
            d = rng.randint(max(2, dlo), dhi)
            if gen.get("remainder", False):
                dividend = rng.randint(lo, hi)
            else:
                qlo, qhi = -(-lo // d), hi // d
                if qlo > qhi:
                    continue
                dividend = d * rng.randint(qlo, qhi)
            q, r = divmod(dividend, d)
            ans = str(q) if r == 0 else f"{q} R {r}"
            return {"key": ("div", dividend, d), "stem": f"{dividend} ÷ {d} =", "answer": ans}

    raise ValueError(f"unknown gen type: {t}")


def build_sheet(gen, count, rng=None):
    """Return `count` problems from a gen spec, avoiding duplicates where the
    problem space allows it, and never repeating the immediately-previous item."""
    rng = rng or random.Random()
    out, seen, last = [], set(), None
    attempts = 0
    while len(out) < count and attempts < count * 60:
        attempts += 1
        p = _one(gen, rng)
        if p["key"] == last:
            continue
        if p["key"] in seen and attempts < count * 30:
            continue
        out.append(p)
        seen.add(p["key"])
        last = p["key"]
    return out


if __name__ == "__main__":
    import json, os
    HERE = os.path.dirname(os.path.abspath(__file__))
    levels = json.load(open(os.path.join(HERE, "..", "data", "drill_levels.json")))
    rng = random.Random(7)
    for strand in levels["strands"]:
        print(f"\n=== {strand['name']} ===")
        for lvl in strand["levels"]:
            sheet = build_sheet(lvl["gen"], 8, rng)
            uniq = len({p["key"] for p in sheet})
            print(f"  {lvl['code']:<8} {lvl['title']:<42} unique {uniq}/8  e.g. {sheet[0]['stem']} {sheet[0]['answer']}")

File: app/lib/domain/drill_gen.dart
import 'dart:math';

/// Procedural problem generator for the speed & accuracy drills — a Dart port
/// of content/tools/drill_gen.py. Kept in lock-step with that file so a
/// worksheet feels the same whether it's printed or answered in the app;
/// the level ladder itself (titles, sequence, these `gen` specs) lives in
/// Postgres (drill_levels), not duplicated here.
class DrillProblem {
  final String key; // for de-duplication within one sheet
  final String stem; // "47 + 26 ="
  final String answer; // "73" or "12 R 2"
  const DrillProblem({required this.key, required this.stem, required this.answer});
}

(int, int) _digitRange(int n) => switch (n) {
      1 => (1, 9),
      2 => (10, 99),
      3 => (100, 999),
      4 => (1000, 9999),
      _ => throw ArgumentError('unsupported digit count: $n'),
    };

int _randInt(Random rng, int lo, int hi) => rng.nextInt(hi - lo + 1) + lo;

bool _hasCarry(int a, int b, int width) {
  final da = a.toString().padLeft(width, '0');
  final db = b.toString().padLeft(width, '0');
  for (var i = 0; i < width; i++) {
    if ((int.parse(da[i]) + int.parse(db[i])) >= 10) return true;
  }
  return false;
}

bool _hasBorrow(int a, int b, int width) {
  final da = a.toString().padLeft(width, '0');
  final db = b.toString().padLeft(width, '0');
  for (var i = 0; i < width; i++) {
    if (int.parse(da[i]) < int.parse(db[i])) return true;
  }
  return false;
}

List<int> _range2(Object o) => (o as List).map((e) => e as int).toList();

DrillProblem _one(Map<String, dynamic> gen, Random rng) {
  final type = gen['type'] as String;

  switch (type) {
    case 'add':
      if (gen.containsKey('a') && gen.containsKey('b')) {
        final ar = _range2(gen['a']), br = _range2(gen['b']);
        final minSum = (gen['min_sum'] as int?) ?? 0;
        final maxSum = (gen['max_sum'] as int?) ?? (1 << 30);
        for (var i = 0; i < 500; i++) {
          final a = _randInt(rng, ar[0], ar[1]);
          final b = _randInt(rng, br[0], br[1]);
          final s = a + b;
          if (s < minSum || s > maxSum) continue;
          return DrillProblem(key: 'add:$a:$b', stem: '$a + $b =', answer: '$s');
        }
      }
      final ra = _digitRange(gen['digits_a'] as int);
      final rb = _digitRange(gen['digits_b'] as int);
      final width = max(gen['digits_a'] as int, gen['digits_b'] as int);
      final wantCarry = gen['carry'] as bool?;
      for (var i = 0; i < 500; i++) {
        final a = _randInt(rng, ra.$1, ra.$2);
        final b = _randInt(rng, rb.$1, rb.$2);
        if (wantCarry != null && _hasCarry(a, b, width) != wantCarry) continue;
        return DrillProblem(key: 'add:$a:$b', stem: '$a + $b =', answer: '${a + b}');
      }
      throw StateError('add: could not satisfy constraints');

    case 'add_column':
      final rd = _digitRange(gen['digits'] as int);
      final count = gen['count'] as int;
      final nums = List.generate(count, (_) => _randInt(rng, rd.$1, rd.$2));
      return DrillProblem(
        key: 'addcol:${nums.join(",")}',
        stem: '${nums.join(" + ")} =',
        answer: '${nums.fold(0, (s, n) => s + n)}',
      );

    case 'sub':
      final ra = _digitRange(gen['digits_a'] as int);
      final rb = _digitRange(gen['digits_b'] as int);
      final width = max(gen['digits_a'] as int, gen['digits_b'] as int);
      final wantBorrow = gen['borrow'] as bool?;
      final needZero = (gen['zeros'] as bool?) ?? false;
      for (var i = 0; i < 800; i++) {
        final a = _randInt(rng, ra.$1, ra.$2);
        final b = _randInt(rng, rb.$1, rb.$2);
        if (a < b) continue;
        if (wantBorrow != null && _hasBorrow(a, b, width) != wantBorrow) continue;
        if (needZero && !a.toString().substring(1).contains('0')) continue;
        return DrillProblem(key: 'sub:$a:$b', stem: '$a − $b =', answer: '${a - b}');
      }
      throw StateError('sub: could not satisfy constraints');

    case 'table':
      {
        final table = gen['table'] as int;
        final r = _range2(gen['range']);
        final n = _randInt(rng, r[0], r[1]);
        final swap = rng.nextDouble() < 0.5;
        final left = swap ? n : table, right = swap ? table : n;
        return DrillProblem(key: 'mul:$table:$n', stem: '$left × $right =', answer: '${table * n}');
      }

    case 'table_mixed':
      {
        final tables = _range2(gen['tables']);
        final table = tables[rng.nextInt(tables.length)];
        final r = _range2(gen['range']);
        final n = _randInt(rng, r[0], r[1]);
        final swap = rng.nextDouble() < 0.5;
        final left = swap ? n : table, right = swap ? table : n;
        return DrillProblem(key: 'mul:$table:$n', stem: '$left × $right =', answer: '${table * n}');
      }

    case 'mul':
      final ra = _digitRange(gen['digits_a'] as int);
      final rb = _digitRange(gen['digits_b'] as int);
      final digitsB = gen['digits_b'] as int;
      final wantCarry = gen['carry'] as bool?;
      for (var i = 0; i < 500; i++) {
        final a = _randInt(rng, ra.$1, ra.$2);
        final b = _randInt(rng, rb.$1, rb.$2);
        if (digitsB == 1 && wantCarry != null) {
          final hasCarry = a.toString().split('').any((d) => (int.parse(d) * b) >= 10);
          if (hasCarry != wantCarry) continue;
        }
        return DrillProblem(key: 'mul:$a:$b', stem: '$a × $b =', answer: '${a * b}');
      }
      throw StateError('mul: could not satisfy constraints');

    case 'div_fact':
      {
        final r = _range2(gen['range']);
        final divisor = _randInt(rng, r[0], r[1]);
        final quotient = _randInt(rng, r[0], r[1]);
        final dividend = divisor * quotient;
        return DrillProblem(
            key: 'div:$dividend:$divisor', stem: '$dividend ÷ $divisor =', answer: '$quotient');
      }

    case 'div':
      final rd = _digitRange(gen['digits_dividend'] as int);
      final dr = _range2(gen['divisor']);
      final wantRemainder = (gen['remainder'] as bool?) ?? false;
      for (var i = 0; i < 500; i++) {
        final d = _randInt(rng, max(2, dr[0]), dr[1]);
        int dividend;
        if (wantRemainder) {
          dividend = _randInt(rng, rd.$1, rd.$2);
        } else {
          final qlo = (rd.$1 + d - 1) ~/ d; // ceil
          final qhi = rd.$2 ~/ d;
          if (qlo > qhi) continue;
          dividend = d * _randInt(rng, qlo, qhi);
        }
        final q = dividend ~/ d, r = dividend % d;
        final ans = r == 0 ? '$q' : '$q R $r';
        return DrillProblem(key: 'div:$dividend:$d', stem: '$dividend ÷ $d =', answer: ans);
      }
      throw StateError('div: could not satisfy constraints');

    default:
      throw ArgumentError('unknown gen type: $type');
  }
}

/// Returns `count` problems from a gen spec, avoiding duplicates where the
/// problem space allows it, and never repeating the immediately-previous item
/// — same rules as the Python generator, so in-app and printed sheets match.
List<DrillProblem> buildSheet(Map<String, dynamic> gen, int count, [Random? rng]) {
  final r = rng ?? Random();
  final out = <DrillProblem>[];
  final seen = <String>{};
  String? last;
  var attempts = 0;
  while (out.length < count && attempts < count * 60) {
    attempts++;
    final p = _one(gen, r);
    if (p.key == last) continue;
    if (seen.contains(p.key) && attempts < count * 30) continue;
    out.add(p);
    seen.add(p.key);
    last = p.key;
  }
  return out;
}
